### DISCUSSION TITLE
Customization point for iterators to be passed directly

### DISCUSSION THREAD

**Discussion by masterleinad** [uxlfoundation/oneDPL#1939] [Ideas]
https://github.com/oneapi-src/oneDPL/issues/855 provides an example were users want to provide their own iterators that reference data on the device. Currently, this causes segfaults since `oneDPL` is creating a CPU-side copy of the respective elements before launching the algorithm.
This problem also appeared in `Kokkos`, see https://github.com/kokkos/kokkos/pull/7502. `Kokkos` defines a `RandomAccessIterator` class that references a `Kokkos::View` that manges the respective data and has similar functionality to `mdspan` (the latter was inspired by the former). This means that the access pattern can be anything although in most cases consecutive elements are a constant stride apart. This means that simply passing raw pointers to `oneDPL` doesn't represent the iterator space correctly and `oneDPL` algorithms can't be used (without copying data to a continuous buffer on the host or device).

Whether to create CPU-side buffers is controlled by `oneapi::dpl::__internal::is_passed_directly` which basically looks for an `is_passed_directly` alias in the iterator class.
This works but pollutes user-defined classes (and the alias name doesn't indicate that the alias is used in `oneDPL`).

https://github.com/oneapi-src/oneDPL/pull/861 provides a solution that provides a wrapper class called `direct_iterator` so that forwards all iterator calls the wrapped class. A different solution is opposed in https://github.com/oneapi-src/oneDPL/pull/861#issuecomment-2444273729 by introducing a customization point in the public `onedpl` namespace such as `onedpl::is_passed_directly` similar to `sycl::is_device_copyable`.

**Comment by masterleinad** [comment]
For `Kokkos`, it doesn't make much of a difference if a wrapper class or a type trait is used. We need to wrap a user-provided comparator anyway to declare it as device-copyable.

**Reply by danhoeflinger** [reply]
Thanks for bringing this up, I agree this would be a good feature to add officially and I think I prefer the public trait similar to `sycl::is_device_copyable`.  

Currently, SYCLomatic helper headers use the existing undocumented option you describe of defining `is_passed_directly` within the iterator classes to accomplish this.  I agree that this is not ideal for the reasons you describe, and is not documented or specified officially.  I think we can use those iterators in SYCLomatic to provide an example of how this can be used as we release this.

I think that the public type trait would fit easily with the internal infrastructure we have built to handle this.  It also should be familiar for users of sycl interested in this sort of thing (`sycl::is_device_copyable`).  

I can try to take a look at mocking this up, if we agree that this is the preferred approach.  

We would need to consider how to handle this as it comes to the specification. I think adding it to the specification would not be a problem, but that is a necessary step, or alternatively this could start as an `extension` or `experimental`.

**Reply by danhoeflinger** [reply]
I should also say that we can create an RFC for this, if we agree generally on the approach here.  I don't expect many surprises there, as there is a good amount of existing infrastructure for which this is a relatively small extension, and also a close example from `sycl::is_device_copyable` which we can model this on.

This RFC can serve as a starting point for future documentation and specification as well.

**Reply by akukanov** [reply]
> I think that the public type trait would fit easily with the internal infrastructure we have built to handle this. It also should be familiar for users of sycl interested in this sort of thing (`sycl::is_device_copyable`).

I also prefer the trait/customization point approach to the wrapper class. But we have seen that `sycl::is_device_copyable` is not quite convenient when it comes to customization, we need a way to customize that is less intrusive.

@danhoeflinger will you create a formal RFC proposal for this?

**Reply by masterleinad** [reply]
> But we have seen that `sycl::is_device_copyable` is not quite convenient when it comes to customization

Can you elaborate on that? My only problems that it was not clear whether `sycl::is_device_copyable<T>` implies `sycl::is_device_copyable<const T>` (and the behavior actually changed between releases).

**Reply by akukanov** [reply]
@masterleinad, the way to customize `sycl::is_device_copyable` is to explicitly provide the specialization for the type needed. That creates kind of an ugly dilemma for internal types defined in a deeply nested namespace: either a) close all the namespaces, specialize the trait, open everything again, or b) move the trait specialization somewhat far away from the class definition. We ended up with (b) combining all such specializations in a single file with some macro machinery to hide boilerplate; but it would be much better experience if it was sufficient to add something to the type definition that would be automatically recognized by `sycl::is_device_copyable`.

**Reply by danhoeflinger** [reply]
Yes, I can make an RFC.  I will have to juggle this with other priorities and some time off coming up, but I will commit to doing this. 

As for the issues with `is_device_copyable`, there have been a few things, but the first being most important by a large factor.  The issue with `const` is frustrating, but I believe Alexey is referring to the following issues we have seen: 

1) From my perspective, the biggest problems/ downsides with `sycl::is_device_copyable` are not so much the usage of the trait, but rather the number of cases which seem that they should be automatically determined to be device copyable, but still require `sycl::is_device_copyable` to be defined.  The problem with sycl device copyable is that any non device copyable type "infects" any template structure which may depend on it as a member variable.  In our case, any internal helper type which may wrap user provided types needs to have a specialization for `sycl::is_device_copyable`. 

2) We have also seen issues with maintenance.  There are structures / classes within oneDPL which are used both for device policies and host policies, and we have not wanted to clutter the implementation of this unified definition with sycl specific code.  We therefore created a separate `sycl_traits.h` header which contains all necessary definitions for device copyability.  However, especially with this separation, maintenance has been somewhat difficult, where it may not be obvious that such a specialization is required to be added or modified when adding or modifying similar structures.

However, in my opinion, this second issue largely goes away when this first issue is dealt with.  If types composed fully of device copyable types are taken care of automatically, then the amount of specializations required shrinks to a very small number, and only to types which do something actively to require device copyability specialization.  This is way easier to maintain when it is more rare, and also more clear that the new changes / structure are actively the entity responsible for the need for device copyable specialization.  
(1+2 is the reason for the recent regression in sort.  Some new changes to improve merge sort performance missed sycl device copyable specialization for intermediate helper types which may wrap user provided types which may be device copyable but not trivially copyable.)

3) `is_device_copyable` was a little awkward to make helper structures / macros to share commonly repeated structure and minimizer redundant code (to handle the cases mentioned above where all template type member variables are device copyable).  See `sycl_traits.h` for an example of this awkwardness. I think this third issue also mostly goes away if (1) goes away, as specialization with redundant infrastructure required largely goes away.  Each case which wouldn't fall into this automatic case wouldn't fall into the category which would rely on this redundant structure, but rather something custom and specific for that case.

To bring this back to `is_passed_directly`,  I think that it will be much less likely to see specializations which depend on some check of template types, and more cases which are just merely defined to be `true_type` by users. However, this perhaps needs to be considered in more detail with example cases.  If anyone has examples of such cases, this would be useful to craft the RFC. 

I don't believe oneDPL has the power to automatically mark types composed fully of `is_passed_directly` to be automatically `is_passed_directly` without an explicit specialization (as the compiler may be able to do with device copyable), so I suppose in that way we would be falling into a similar trap.  However, I don't expect as much infrastructure to be built upon this trait to cause similar issues.  We already have the mechanisms to allow `is_passed_directly` to be applied to `zip_iterator` which are composed fully of `is_passed_directly` iterators.  The same for the rest of our iterator types which wrap other iterators.

**Reply by danhoeflinger** [reply]
> but it would be much better experience if it was sufficient to add something to the type definition that would be automatically recognized by `sycl::is_device_copyable`.

To some extent we have this already.  It is the `is_passed_directly` trait defined in the user's type we call the legacy "is_passed_directly" trait (What SYCLomatic currently uses).  If we wanted to add a different one to include `ONEDPL` in the name or something to make it more clear, we could add this into the spec / docs, and add it to our infrastructure.

**Reply by masterleinad** [reply]
I would be opposed to require an alias (as it is done right now) since it pollutes the respective class with information that is only relevant for interoperability with `oneDPL`.

**Reply by danhoeflinger** [reply]
I don't see a reason we could not support **both** options, a public specialization similar to `sycl::is_device_copyable` and an alias in the user's type.  If either is defined to `true_type` then we could treat it accordingly.  Users could pick what works best for them.  @masterleinad, In your case you could ignore that it exists, and use the `oneapi::dpl::is_passed_directly` specialization for your type.

**Reply by masterleinad** [reply]
>  If either is defined to true_type then we could treat it accordingly.

Sure, it just needs to be clear (and tested) what happens in all four cases.

**Reply by danhoeflinger** [reply]
I agree, it needs to be clearly specified and tested.  

If we choose to instead just go with one, at this point, I'd prefer the `oneapi::dpl::is_passed_directly` specializations because I don't think the downsides will be as severe as see with `sycl::is_device_copyable` for the reasons discussed above.

**Comment by danhoeflinger** [comment]
@akukanov  @masterleinad
I've provided an initial draft here #1999.  I think it seems to be a good fit for what we have described here.  I think its best to go for a single solution rather than offer multiple options.  After some investigation, I agree with @akukanov that the customization point route is the cleanest option as compared to the other options floated.

**Comment by akukanov** [comment]
The specification patch: https://github.com/uxlfoundation/oneAPI-spec/pull/620
The implementation: https://github.com/uxlfoundation/oneDPL/pull/2126

**Comment by akukanov** [comment]
Implemented in https://github.com/uxlfoundation/oneDPL/releases/tag/oneDPL-2022.9.0-release
